8263506: Make sun.net.httpserver.UnmodifiableHeaders unmodifiable

FrauBoes opened the pull request:
The fix makes the map in sun.net.httpserver.UnmodifiableHeaders unmodifiable by wrapping it in an unmodifiable view.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263506](https://bugs.openjdk.java.net/browse/JDK-8263506): Make sun.net.httpserver.UnmodifiableHeaders unmodifiable


### Reviewers
 * [Michael McMahon](https://openjdk.java.net/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to 8981b44400727984bce66b99d9bed9cb6baa0e0c
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3032/head:pull/3032` \
`$ git checkout pull/3032`

Update a local copy of the PR: \
`$ git checkout pull/3032` \
`$ git pull https://git.openjdk.java.net/jdk pull/3032/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3032`

View PR using the GUI difftool: \
`$ git pr show -t 3032`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3032.diff">https://git.openjdk.java.net/jdk/pull/3032.diff</a>

</details>

Michael-Mc-Mahon commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -52,11 +53,12 @@ public boolean containsValue(Object value) {
+            final var headers = new Headers();
+            map.forEach((k, v) -> headers.add(k, v.get(0)));
+            return headers.getFirst(key);
         }
```
Should be possible to just return a reference to the first String in the map rather than instantiating a new Headers and populating it just for the duration of the call.

FrauBoes replied:
Headers normalizes the key before any get/getFirst, for example Headers line 131:

`public List<String> get(Object key) {
            return map.get(normalize((String)key));
        }`

I assume we want to maintain this behaviour, but there might be a better way than creating this local Headers?

Michael-Mc-Mahon replied:
Actually, I just noticed that the List<String>'s are still modifiable. The constructor might have to create an intermediate Map<String,List<String>> with unmodifiable Lists and then you create an unmodifiable Map from that. 

As regards the comment above, yes you want to maintain that behavior. Could you normalize the key here and then just return `map.get(key).get(0)` (with suitable null checks)?

dfuch replied:
Argh. Good catch Michael.  So now we'd also have to reimplement valueSet() and get() to wrap things into unmodifiableList... Or could we get away with providing an immutable snapshot instead of an immutable view?

FrauBoes replied:
Thanks for catching that. The `List<String>` is now wrapped in an unmodifiable map, and some cases were added to the test.

Michael-Mc-Mahon commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -82,7 +84,7 @@ public void clear() {
         public Set<String> keySet() {
-            return Collections.unmodifiableSet (map.keySet());
+            return Collections.unmodifiableSet(map.keySet());
         }
```
Should be able to just return "map.keySet()" here because map is already unmodifiable

Michael-Mc-Mahon commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -92,7 +94,19 @@ public void clear() {
         public Set<Map.Entry<String, List<String>>> entrySet() {
-            return Collections.unmodifiableSet (map.entrySet());
+            return Collections.unmodifiableSet(map.entrySet());
+        }
```
Same as previous comment.

dfuch commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -26,14 +26,15 @@
+public class UnmodifiableHeaders extends Headers {
 
-        Headers map;
+        private Map<String, List<String>> map;
```
How about:

```
private Headers headers;
private Map<String, List<String>> map;

public UnmodifiableHeaders(Headers map) {
    this.map = Collections.unmodifiableMap(map);
    this.headers = headers;
}
```

Then getFirst simply becomes `return headers.getFirst(key);`

Michael-Mc-Mahon replied:
I guess this presumes that no reference to `headers` escapes? Can that be assumed?

dfuch replied:
Yes - it's private - and that wasn't a problem before - was it?

FrauBoes replied:
Added (back) a `headers` field as per you suggestion, which is only accessed by `getFirst`.

dfuch commented:
The changes LGTM. What about modifying ExchangeImpl to create the UnmodifiableHeaders in its constructor, instead of creating a new instance every time getRequestHeaders() is called?

Michael-Mc-Mahon commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -26,15 +26,21 @@
+public class UnmodifiableHeaders extends Headers {
 
-        Headers map;
+    private final Headers headers;
```
Could you add a comment here that Headers is mutable, but no reference to it escapes.

Also, can you fix the indentation of the other existing methods.

dfuch reviewed (approved):
LGTM

Michael-Mc-Mahon reviewed (approved):
LGTM

FrauBoes commented:
I included a null check in the constructor to account for null values of `List<String>`, which `Headers` allows. Fixed the indentation.

Testing: Tier 1-3 all clear.

dfuch commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -31,24 +31,27 @@
-        this.map = Collections.unmodifiableMap(unmodListMap);
+        var unmodHeaders = new Headers();
+        h.forEach((k, v) -> {
+                List<String> l = v == null ? List.of() : Collections.unmodifiableList(v);
```
If v is `null` it should be mapped to `null` - not to `List.of()`, to preserve map equality.

dfuch commented on test/jdk/com/sun/net/httpserver/UnmodifiableHeadersTest.java:
```diff
@@ -88,6 +88,7 @@ static void assertUnmodifiableCollection(Headers headers) {
 
     static void assertUnmodifiableList(Headers headers) {
```
Maybe the test should be extended to double check that:

```
var headers = ...
var unmodifiableHeaders = new UnmodifiableHeaders(headers);
assertEquals(unmodifiableHeaders, headers);
assertEquals(unmodifiableHeaders.hashCode(), headers.hashCode());
```

FrauBoes commented:
It turns out I wasn't right in the last commit. The null check is not needed in the constructor, but in the `get` method, where `Headers::get` can return null, which can cause `Collections.unmodifiableList` to throw NPE. I updated the get method and added some more test cases to confirm equality of Headers and UnmodifiableHeaders instances.

Testing: Tier 1-3 all clear.

dfuch commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -53,7 +50,10 @@ public UnmodifiableHeaders(Headers headers) {
     public boolean containsValue(Object value) { return headers.containsValue(value); }
 
-    public List<String> get(Object key) { return map.get(key); }
+    public List<String> get(Object key) {
+        List<String> l = headers.get(key);
+        return l == null ? null : Collections.unmodifiableList(l);
+    }
```
You could simply return `map.get(k)` here - since map is an unmodifiable map wrapping  `unmodHeaders`

dfuch commented on src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java:
```diff
@@ -37,10 +37,7 @@
-        });
+        h.forEach((k, v) -> unmodHeaders.put(k, Collections.unmodifiableList(v)));
         this.map = Collections.unmodifiableMap(unmodHeaders);
         this.headers = h;
```
Now that we have `unmodHeaders` we could simply keep it:

```
this.headers = unmodHeaders;
```

we no longer need to keep the original `headers` around.

Michael-Mc-Mahon commented:
Latest changes LGTM.

FrauBoes commented:
/integrate
